File: muscle-and-fitness-server-spring-boot-starter/src/main/java/com/jmsoftware/maf/muscleandfitnessserverspringbootstarter/helper/IpHelper.java
package com.jmsoftware.maf.muscleandfitnessserverspringbootstarter.helper;

import cn.hutool.core.util.StrUtil;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.boot.web.context.WebServerInitializedEvent;
import org.springframework.context.ApplicationListener;
import org.springframework.core.env.Environment;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.net.URL;
import java.util.Enumeration;

/**
 * <h1>ServerConfiguration</h1>
 * <p>
 * Change description here.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com
 * @date 2019-04-26 16:02
 **/
@Slf4j
@Getter
@Component
@RequiredArgsConstructor
public class IpHelper implements ApplicationListener<WebServerInitializedEvent> {
    private static final String DEVELOPMENT_ENVIRONMENT = "development";
    @Getter(AccessLevel.NONE)
    private final Environment environment;
    private int serverPort;

    @Override
    public void onApplicationEvent(WebServerInitializedEvent event) {
        this.serverPort = event.getWebServer().getPort();
    }

    /**
     * Find public IP address.
     *
     * @return public IP
     */
    public String getPublicIp() {
        String jointProfiles = String.join(",", environment.getActiveProfiles());
        if (StrUtil.isNotBlank(jointProfiles)) {
            if (jointProfiles.contains(DEVELOPMENT_ENVIRONMENT)) {
                log.debug("Current active profiles for environment contains: {}", DEVELOPMENT_ENVIRONMENT);
                return this.getInternetIp();
            }
        }
        try {
            // An API provided by https://whatismyipaddress.com/api
            val url = new URL("https://ipv4bot.whatismyipaddress.com/");
            val sc = new BufferedReader(new InputStreamReader(url.openStream()));
            // Read system IP Address
            return sc.readLine().trim();
        } catch (Exception e) {
            log.error("Cannot execute properly to get IP address from https://whatismyipaddress.com/api", e);
        }
        return this.getInternetIp();
    }

    /**
     * Get internet IP.
     *
     * @return internet IP
     */
    private String getInternetIp() {
        val intranetIp = this.getIntranetIp();
        try {
            val networks = NetworkInterface.getNetworkInterfaces();
            InetAddress ip;
            Enumeration<InetAddress> addresses;
            while (networks.hasMoreElements()) {
                addresses = networks.nextElement().getInetAddresses();
                while (addresses.hasMoreElements()) {
                    ip = addresses.nextElement();
                    if (ip instanceof Inet4Address
                            && ip.isSiteLocalAddress()
                            && !ip.getHostAddress().equals(intranetIp)) {
                        return ip.getHostAddress();
                    }
                }
            }
            return intranetIp;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Get intranet IP.
     *
     * @return intranet IP
     */
    private String getIntranetIp() {
        try {
            return InetAddress.getLocalHost().getHostAddress();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: reactive-spring-cloud-starter/src/main/java/com/jmsoftware/maf/reactivespringcloudstarter/helper/IpHelper.kt
package com.jmsoftware.maf.reactivespringcloudstarter.helper

import cn.hutool.core.util.StrUtil
import cn.hutool.http.HttpUtil
import com.jmsoftware.maf.common.util.logger
import org.springframework.boot.web.context.WebServerInitializedEvent
import org.springframework.context.ApplicationListener
import org.springframework.core.env.Environment
import java.net.Inet4Address
import java.net.InetAddress
import java.net.NetworkInterface
import java.util.*

/**
 * # IpHelper
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/17/22 8:05 AM
 */
class IpHelper(
    private val environment: Environment
) : ApplicationListener<WebServerInitializedEvent> {
    companion object {
        private const val DEVELOPMENT_ENVIRONMENT = "development"
        private val log = logger()
    }

    var serverPort = 0

    override fun onApplicationEvent(event: WebServerInitializedEvent) {
        serverPort = event.webServer.port
    }

    /**
     * Find public IP address.
     *
     * @return public IP
     */
    fun getPublicIp(): String {
        val jointProfiles = java.lang.String.join(",", *environment.activeProfiles)
        if (StrUtil.containsIgnoreCase(jointProfiles, DEVELOPMENT_ENVIRONMENT)) {
            log.debug("Current active profiles for environment contains: $DEVELOPMENT_ENVIRONMENT")
            return internetIp()
        }
        // An API provided by https://whatismyipaddress.com/api
        return try {
            HttpUtil.get("https://ipv4bot.whatismyipaddress.com/").trim { it <= ' ' }
        } catch (e: Exception) {
            log.warn("Failed to get public IP address, fallback to intranet IP address. Exception: ${e.message}")
            internetIp()
        }
    }

    /**
     * Get internet IP.
     *
     * @return internet IP
     */
    private fun internetIp(): String {
        val intranetIp = intranetIp()
        val networks = NetworkInterface.getNetworkInterfaces()
        var ip: InetAddress
        var addresses: Enumeration<InetAddress>
        while (networks.hasMoreElements()) {
            addresses = networks.nextElement().inetAddresses
            while (addresses.hasMoreElements()) {
                ip = addresses.nextElement()
                if (ip is Inet4Address
                    && ip.isSiteLocalAddress()
                    && ip.getHostAddress() != intranetIp
                ) {
                    return ip.getHostAddress()
                }
            }
        }
        return intranetIp
    }

    /**
     * Get intranet IP.
     *
     * @return intranet IP
     */
    private fun intranetIp(): String = InetAddress.getLocalHost().hostAddress
}
